syntax = "proto3";

package raft;

// Service định nghĩa các RPC methods cho RAFT
service RaftService {
  // Leader election
  rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
  
  // Log replication
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
  
  // Client operations
  rpc ClientRequest(ClientRequestMsg) returns (ClientResponseMsg);
  
  // Network partition simulation
  rpc DisconnectNodes(DisconnectRequest) returns (DisconnectResponse);
  
  // Get node status
  rpc GetStatus(StatusRequest) returns (StatusResponse);
}

// RequestVote RPC
message RequestVoteRequest {
  int32 term = 1;              // Candidate's term
  string candidate_id = 2;      // Candidate requesting vote
  int32 last_log_index = 3;     // Index of candidate's last log entry
  int32 last_log_term = 4;      // Term of candidate's last log entry
}

message RequestVoteResponse {
  int32 term = 1;               // Current term, for candidate to update itself
  bool vote_granted = 2;        // True means candidate received vote
  string voter_id = 3;          // ID of the voter
}

// AppendEntries RPC (heartbeat và log replication)
message AppendEntriesRequest {
  int32 term = 1;               // Leader's term
  string leader_id = 2;         // Leader's ID
  int32 prev_log_index = 3;     // Index of log entry immediately preceding new ones
  int32 prev_log_term = 4;      // Term of prev_log_index entry
  repeated LogEntry entries = 5; // Log entries to store (empty for heartbeat)
  int32 leader_commit = 6;      // Leader's commit index
}

message AppendEntriesResponse {
  int32 term = 1;               // Current term, for leader to update itself
  bool success = 2;             // True if follower contained entry matching prev_log_index and prev_log_term
  int32 match_index = 3;        // Index of highest log entry known to be replicated
  string follower_id = 4;       // ID of the follower
}

// Log entry structure
message LogEntry {
  int32 term = 1;               // Term when entry was received by leader
  int32 index = 2;              // Log index
  string command = 3;           // Command for state machine (e.g., "SET key value")
  string client_id = 4;         // Client ID that issued the command
}

// Client request
message ClientRequestMsg {
  string command = 1;           // Command to execute (e.g., "SET key value", "GET key", "DELETE key")
  string client_id = 2;         // Client ID
}

message ClientResponseMsg {
  bool success = 1;             // Whether request was successful
  string result = 2;            // Result of the operation
  string leader_id = 3;         // Current leader ID (for redirect if needed)
  string error = 4;             // Error message if any
}

// Network partition simulation
message DisconnectRequest {
  repeated string node_ids = 1; // List of node IDs to disconnect from
}

message DisconnectResponse {
  bool success = 1;
  string message = 2;
}


// Status request
message StatusRequest {
  string requesting_node_id = 1;
}

message StatusResponse {
  string node_id = 1;
  string state = 2;             // Current state: FOLLOWER, CANDIDATE, LEADER
  int32 current_term = 3;
  string voted_for = 4;         // ID of node voted for in current term
  int32 commit_index = 5;
  int32 last_applied = 6;
  int32 log_length = 7;
  repeated string cluster_nodes = 8; // List of all nodes in cluster
  string leader_id = 9;
  map<string, string> state_machine = 10; // Current key-value store
}